api/operator/v1alpha1/istiocsr_types.go

```diff
@@ -0,0 +1,246 @@
+	// +optional
+	SignatureAlgorithm string `json:"signatureAlgorithm,omitempty"`
+
+	// MaxCertificateDuration is the maximum validity duration that can be
+	// requested for a certificate.
+	// +kubebuilder:default:="1h"
+	// +optional
+	MaxCertificateDuration *metav1.Duration `json:"maxCertificateDuration,omitempty"`
+}
+
+// ServerConfig is for configuring the server endpoint used by istio
+// for obtaining the certificates.
+type ServerConfig struct {
+	// port to serve istio-csr gRPC service.
+	// +kubebuilder:default:=443
+	// +kubebuilder:validation:XValidation:rule="oldSelf == 0 || self == oldSelf",message="port is immutable once set"
+	// +optional
+	Port int32 `json:"port,omitempty"`
+}
+
+// IstioConfig is for configuring the istio specifics.
+type IstioConfig struct {
+	// revisions are the istio revisions that are currently installed in the cluster.
+	// Changing this field will modify the DNS names that will be requested for
+	// the istiod certificate.
+	// +listType=atomic
+	// +kubebuilder:default:={"default"}
+	// +kubebuilder:validation:XValidation:rule="self.all(x, x in oldSelf) && oldSelf.all(x, x in self)",message="revisions is immutable once set"
+	// +kubebuilder:validation:MaxItems=10
+	// +optional
+	Revisions []string `json:"revisions,omitempty"`
+
+	// namespace of the istio control-plane.
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="namespace is immutable once set"
+	// +required
+	Namespace string `json:"namespace,omitempty"`
+}
+
+// ControllerConfig is for configuring the controller for setting up
+// defaults to enable istio-csr agent.
+type ControllerConfig struct {
+	// labels to apply to all resources created for istio-csr agent deployment.
+	// +mapType=granular
+	// +optional
+	Labels map[string]string `json:"labels,omitempty"`
+}
+
+// IstioCSRStatus is the most recently observed status of the IstioCSR.
+type IstioCSRStatus struct {
+	// conditions holds information of the current state of the istio-csr agent deployment.
+	ConditionalStatus `json:",inline,omitempty"`
+
+	// istioCSRImage is the name of the image and the tag used for deploying istio-csr.
+	IstioCSRImage string `json:"istioCSRImage,omitempty"`
+
+	// istioCSRGRPCEndpoint is the service endpoint of istio-csr made available for user
+	// to configure the same in istiod config to enable istio to use istio-csr for
+	// certificate requests.
+	IstioCSRGRPCEndpoint string `json:"istioCSRGRPCEndpoint,omitempty"`
+
+	// serviceAccount created by the controller for the istio-csr agent.
+	ServiceAccount string `json:"serviceAccount,omitempty"`
+
+	// clusterRoleBinding created by the controller for the istio-csr agent.
+	ClusterRoleBinding string `json:"clusterRoleBinding,omitempty"`
+}
+
+type ConditionalStatus struct {
+	// conditions holds information of the current state of the istio-csr agent deployment.
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
```

Comment:
```suggestion
	// conditions holds information of the current state of the istio-csr agent deployment.
	//
	// +patchMergeKey=type
        // +patchStrategy=merge
        // +listType=map
        // +listMapKey=type
	Conditions []metav1.Condition `json:"conditions,omitempty"`
```

referring to https://github.com/kubernetes/apimachinery/blob/3e8e52d6a1259ada73f63c1c7d1fad39d4ba9fb4/pkg/apis/meta/v1/types.go#L1589-L1604, our kube generators may not have been bumped in a while, however am aware that recent API generators enforce not having `+listType` for slices: better to be consistent.

Reply:
suggestion incorporated.